<html>
<head>
<title>test_asyncgen.py</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8">
<style type="text/css">
.s0 { color: #cc7832;}
.s1 { color: #a9b7c6;}
.s2 { color: #6897bb;}
.s3 { color: #6a8759;}
.s4 { color: #808080;}
</style>
</head>
<body bgcolor="#2b2b2b">
<table CELLSPACING=0 CELLPADDING=5 COLS=1 WIDTH="100%" BGCOLOR="#606060" >
<tr><td><center>
<font face="Arial, Helvetica" color="#000000">
test_asyncgen.py</font>
</center></td></tr></table>
<pre><span class="s0">import </span><span class="s1">sys</span>
<span class="s0">import </span><span class="s1">weakref</span>
<span class="s0">import </span><span class="s1">pytest</span>
<span class="s0">from </span><span class="s1">math </span><span class="s0">import </span><span class="s1">inf</span>
<span class="s0">from </span><span class="s1">functools </span><span class="s0">import </span><span class="s1">partial</span>
<span class="s0">from </span><span class="s1">async_generator </span><span class="s0">import </span><span class="s1">aclosing</span>
<span class="s0">from </span><span class="s1">... </span><span class="s0">import </span><span class="s1">_core</span>
<span class="s0">from </span><span class="s1">.tutil </span><span class="s0">import </span><span class="s1">gc_collect_harder</span><span class="s0">, </span><span class="s1">buggy_pypy_asyncgens</span><span class="s0">, </span><span class="s1">restore_unraisablehook</span>


<span class="s0">def </span><span class="s1">test_asyncgen_basics():</span>
    <span class="s1">collected = []</span>

    <span class="s0">async def </span><span class="s1">example(cause):</span>
        <span class="s0">try</span><span class="s1">:</span>
            <span class="s0">try</span><span class="s1">:</span>
                <span class="s0">yield </span><span class="s2">42</span>
            <span class="s0">except </span><span class="s1">GeneratorExit:</span>
                <span class="s0">pass</span>
            <span class="s0">await </span><span class="s1">_core.checkpoint()</span>
        <span class="s0">except </span><span class="s1">_core.Cancelled:</span>
            <span class="s0">assert </span><span class="s3">&quot;exhausted&quot; </span><span class="s0">not in </span><span class="s1">cause</span>
            <span class="s1">task_name = _core.current_task().name</span>
            <span class="s0">assert </span><span class="s1">cause </span><span class="s0">in </span><span class="s1">task_name </span><span class="s0">or </span><span class="s1">task_name == </span><span class="s3">&quot;&lt;init&gt;&quot;</span>
            <span class="s0">assert </span><span class="s1">_core.current_effective_deadline() == -inf</span>
            <span class="s0">with </span><span class="s1">pytest.raises(_core.Cancelled):</span>
                <span class="s0">await </span><span class="s1">_core.checkpoint()</span>
            <span class="s1">collected.append(cause)</span>
        <span class="s0">else</span><span class="s1">:</span>
            <span class="s0">assert </span><span class="s3">&quot;async_main&quot; </span><span class="s0">in </span><span class="s1">_core.current_task().name</span>
            <span class="s0">assert </span><span class="s3">&quot;exhausted&quot; </span><span class="s0">in </span><span class="s1">cause</span>
            <span class="s0">assert </span><span class="s1">_core.current_effective_deadline() == inf</span>
            <span class="s0">await </span><span class="s1">_core.checkpoint()</span>
            <span class="s1">collected.append(cause)</span>

    <span class="s1">saved = []</span>

    <span class="s0">async def </span><span class="s1">async_main():</span>
        <span class="s4"># GC'ed before exhausted</span>
        <span class="s0">with </span><span class="s1">pytest.warns(</span>
            <span class="s1">ResourceWarning</span><span class="s0">, </span><span class="s1">match=</span><span class="s3">&quot;Async generator.*collected before.*exhausted&quot;</span>
        <span class="s1">):</span>
            <span class="s0">assert </span><span class="s2">42 </span><span class="s1">== </span><span class="s0">await </span><span class="s1">example(</span><span class="s3">&quot;abandoned&quot;</span><span class="s1">).asend(</span><span class="s0">None</span><span class="s1">)</span>
            <span class="s1">gc_collect_harder()</span>
        <span class="s0">await </span><span class="s1">_core.wait_all_tasks_blocked()</span>
        <span class="s0">assert </span><span class="s1">collected.pop() == </span><span class="s3">&quot;abandoned&quot;</span>

        <span class="s4"># aclosing() ensures it's cleaned up at point of use</span>
        <span class="s0">async with </span><span class="s1">aclosing(example(</span><span class="s3">&quot;exhausted 1&quot;</span><span class="s1">)) </span><span class="s0">as </span><span class="s1">aiter:</span>
            <span class="s0">assert </span><span class="s2">42 </span><span class="s1">== </span><span class="s0">await </span><span class="s1">aiter.asend(</span><span class="s0">None</span><span class="s1">)</span>
        <span class="s0">assert </span><span class="s1">collected.pop() == </span><span class="s3">&quot;exhausted 1&quot;</span>

        <span class="s4"># Also fine if you exhaust it at point of use</span>
        <span class="s0">async for </span><span class="s1">val </span><span class="s0">in </span><span class="s1">example(</span><span class="s3">&quot;exhausted 2&quot;</span><span class="s1">):</span>
            <span class="s0">assert </span><span class="s1">val == </span><span class="s2">42</span>
        <span class="s0">assert </span><span class="s1">collected.pop() == </span><span class="s3">&quot;exhausted 2&quot;</span>

        <span class="s1">gc_collect_harder()</span>

        <span class="s4"># No problems saving the geniter when using either of these patterns</span>
        <span class="s0">async with </span><span class="s1">aclosing(example(</span><span class="s3">&quot;exhausted 3&quot;</span><span class="s1">)) </span><span class="s0">as </span><span class="s1">aiter:</span>
            <span class="s1">saved.append(aiter)</span>
            <span class="s0">assert </span><span class="s2">42 </span><span class="s1">== </span><span class="s0">await </span><span class="s1">aiter.asend(</span><span class="s0">None</span><span class="s1">)</span>
        <span class="s0">assert </span><span class="s1">collected.pop() == </span><span class="s3">&quot;exhausted 3&quot;</span>

        <span class="s4"># Also fine if you exhaust it at point of use</span>
        <span class="s1">saved.append(example(</span><span class="s3">&quot;exhausted 4&quot;</span><span class="s1">))</span>
        <span class="s0">async for </span><span class="s1">val </span><span class="s0">in </span><span class="s1">saved[-</span><span class="s2">1</span><span class="s1">]:</span>
            <span class="s0">assert </span><span class="s1">val == </span><span class="s2">42</span>
        <span class="s0">assert </span><span class="s1">collected.pop() == </span><span class="s3">&quot;exhausted 4&quot;</span>

        <span class="s4"># Leave one referenced-but-unexhausted and make sure it gets cleaned up</span>
        <span class="s0">if </span><span class="s1">buggy_pypy_asyncgens:</span>
            <span class="s1">collected.append(</span><span class="s3">&quot;outlived run&quot;</span><span class="s1">)</span>
        <span class="s0">else</span><span class="s1">:</span>
            <span class="s1">saved.append(example(</span><span class="s3">&quot;outlived run&quot;</span><span class="s1">))</span>
            <span class="s0">assert </span><span class="s2">42 </span><span class="s1">== </span><span class="s0">await </span><span class="s1">saved[-</span><span class="s2">1</span><span class="s1">].asend(</span><span class="s0">None</span><span class="s1">)</span>
            <span class="s0">assert </span><span class="s1">collected == []</span>

    <span class="s1">_core.run(async_main)</span>
    <span class="s0">assert </span><span class="s1">collected.pop() == </span><span class="s3">&quot;outlived run&quot;</span>
    <span class="s0">for </span><span class="s1">agen </span><span class="s0">in </span><span class="s1">saved:</span>
        <span class="s0">assert </span><span class="s1">agen.ag_frame </span><span class="s0">is None  </span><span class="s4"># all should now be exhausted</span>


<span class="s0">async def </span><span class="s1">test_asyncgen_throws_during_finalization(caplog):</span>
    <span class="s1">record = []</span>

    <span class="s0">async def </span><span class="s1">agen():</span>
        <span class="s0">try</span><span class="s1">:</span>
            <span class="s0">yield </span><span class="s2">1</span>
        <span class="s0">finally</span><span class="s1">:</span>
            <span class="s0">await </span><span class="s1">_core.cancel_shielded_checkpoint()</span>
            <span class="s1">record.append(</span><span class="s3">&quot;crashing&quot;</span><span class="s1">)</span>
            <span class="s0">raise </span><span class="s1">ValueError(</span><span class="s3">&quot;oops&quot;</span><span class="s1">)</span>

    <span class="s0">with </span><span class="s1">restore_unraisablehook():</span>
        <span class="s0">await </span><span class="s1">agen().asend(</span><span class="s0">None</span><span class="s1">)</span>
        <span class="s1">gc_collect_harder()</span>
    <span class="s0">await </span><span class="s1">_core.wait_all_tasks_blocked()</span>
    <span class="s0">assert </span><span class="s1">record == [</span><span class="s3">&quot;crashing&quot;</span><span class="s1">]</span>
    <span class="s1">exc_type</span><span class="s0">, </span><span class="s1">exc_value</span><span class="s0">, </span><span class="s1">exc_traceback = caplog.records[</span><span class="s2">0</span><span class="s1">].exc_info</span>
    <span class="s0">assert </span><span class="s1">exc_type </span><span class="s0">is </span><span class="s1">ValueError</span>
    <span class="s0">assert </span><span class="s1">str(exc_value) == </span><span class="s3">&quot;oops&quot;</span>
    <span class="s0">assert </span><span class="s3">&quot;during finalization of async generator&quot; </span><span class="s0">in </span><span class="s1">caplog.records[</span><span class="s2">0</span><span class="s1">].message</span>


<span class="s1">@pytest.mark.skipif(buggy_pypy_asyncgens</span><span class="s0">, </span><span class="s1">reason=</span><span class="s3">&quot;pypy 7.2.0 is buggy&quot;</span><span class="s1">)</span>
<span class="s0">def </span><span class="s1">test_firstiter_after_closing():</span>
    <span class="s1">saved = []</span>
    <span class="s1">record = []</span>

    <span class="s0">async def </span><span class="s1">funky_agen():</span>
        <span class="s0">try</span><span class="s1">:</span>
            <span class="s0">yield </span><span class="s2">1</span>
        <span class="s0">except </span><span class="s1">GeneratorExit:</span>
            <span class="s1">record.append(</span><span class="s3">&quot;cleanup 1&quot;</span><span class="s1">)</span>
            <span class="s0">raise</span>
        <span class="s0">try</span><span class="s1">:</span>
            <span class="s0">yield </span><span class="s2">2</span>
        <span class="s0">finally</span><span class="s1">:</span>
            <span class="s1">record.append(</span><span class="s3">&quot;cleanup 2&quot;</span><span class="s1">)</span>
            <span class="s0">await </span><span class="s1">funky_agen().asend(</span><span class="s0">None</span><span class="s1">)</span>

    <span class="s0">async def </span><span class="s1">async_main():</span>
        <span class="s1">aiter = funky_agen()</span>
        <span class="s1">saved.append(aiter)</span>
        <span class="s0">assert </span><span class="s2">1 </span><span class="s1">== </span><span class="s0">await </span><span class="s1">aiter.asend(</span><span class="s0">None</span><span class="s1">)</span>
        <span class="s0">assert </span><span class="s2">2 </span><span class="s1">== </span><span class="s0">await </span><span class="s1">aiter.asend(</span><span class="s0">None</span><span class="s1">)</span>

    <span class="s1">_core.run(async_main)</span>
    <span class="s0">assert </span><span class="s1">record == [</span><span class="s3">&quot;cleanup 2&quot;</span><span class="s0">, </span><span class="s3">&quot;cleanup 1&quot;</span><span class="s1">]</span>


<span class="s1">@pytest.mark.skipif(buggy_pypy_asyncgens</span><span class="s0">, </span><span class="s1">reason=</span><span class="s3">&quot;pypy 7.2.0 is buggy&quot;</span><span class="s1">)</span>
<span class="s0">def </span><span class="s1">test_interdependent_asyncgen_cleanup_order():</span>
    <span class="s1">saved = []</span>
    <span class="s1">record = []</span>

    <span class="s0">async def </span><span class="s1">innermost():</span>
        <span class="s0">try</span><span class="s1">:</span>
            <span class="s0">yield </span><span class="s2">1</span>
        <span class="s0">finally</span><span class="s1">:</span>
            <span class="s0">await </span><span class="s1">_core.cancel_shielded_checkpoint()</span>
            <span class="s1">record.append(</span><span class="s3">&quot;innermost&quot;</span><span class="s1">)</span>

    <span class="s0">async def </span><span class="s1">agen(label</span><span class="s0">, </span><span class="s1">inner):</span>
        <span class="s0">try</span><span class="s1">:</span>
            <span class="s0">yield await </span><span class="s1">inner.asend(</span><span class="s0">None</span><span class="s1">)</span>
        <span class="s0">finally</span><span class="s1">:</span>
            <span class="s4"># Either `inner` has already been cleaned up, or</span>
            <span class="s4"># we're about to exhaust it. Either way, we wind</span>
            <span class="s4"># up with `record` containing the labels in</span>
            <span class="s4"># innermost-to-outermost order.</span>
            <span class="s0">with </span><span class="s1">pytest.raises(StopAsyncIteration):</span>
                <span class="s0">await </span><span class="s1">inner.asend(</span><span class="s0">None</span><span class="s1">)</span>
            <span class="s1">record.append(label)</span>

    <span class="s0">async def </span><span class="s1">async_main():</span>
        <span class="s4"># This makes a chain of 101 interdependent asyncgens:</span>
        <span class="s4"># agen(99)'s cleanup will iterate agen(98)'s will iterate</span>
        <span class="s4"># ... agen(0)'s will iterate innermost()'s</span>
        <span class="s1">ag_chain = innermost()</span>
        <span class="s0">for </span><span class="s1">idx </span><span class="s0">in </span><span class="s1">range(</span><span class="s2">100</span><span class="s1">):</span>
            <span class="s1">ag_chain = agen(idx</span><span class="s0">, </span><span class="s1">ag_chain)</span>
        <span class="s1">saved.append(ag_chain)</span>
        <span class="s0">assert </span><span class="s2">1 </span><span class="s1">== </span><span class="s0">await </span><span class="s1">ag_chain.asend(</span><span class="s0">None</span><span class="s1">)</span>
        <span class="s0">assert </span><span class="s1">record == []</span>

    <span class="s1">_core.run(async_main)</span>
    <span class="s0">assert </span><span class="s1">record == [</span><span class="s3">&quot;innermost&quot;</span><span class="s1">] + list(range(</span><span class="s2">100</span><span class="s1">))</span>


<span class="s1">@restore_unraisablehook()</span>
<span class="s0">def </span><span class="s1">test_last_minute_gc_edge_case():</span>
    <span class="s1">saved = []</span>
    <span class="s1">record = []</span>
    <span class="s1">needs_retry = </span><span class="s0">True</span>

    <span class="s0">async def </span><span class="s1">agen():</span>
        <span class="s0">try</span><span class="s1">:</span>
            <span class="s0">yield </span><span class="s2">1</span>
        <span class="s0">finally</span><span class="s1">:</span>
            <span class="s1">record.append(</span><span class="s3">&quot;cleaned up&quot;</span><span class="s1">)</span>

    <span class="s0">def </span><span class="s1">collect_at_opportune_moment(token):</span>
        <span class="s1">runner = _core._run.GLOBAL_RUN_CONTEXT.runner</span>
        <span class="s0">if </span><span class="s1">runner.system_nursery._closed </span><span class="s0">and </span><span class="s1">isinstance(</span>
            <span class="s1">runner.asyncgens.alive</span><span class="s0">, </span><span class="s1">weakref.WeakSet</span>
        <span class="s1">):</span>
            <span class="s1">saved.clear()</span>
            <span class="s1">record.append(</span><span class="s3">&quot;final collection&quot;</span><span class="s1">)</span>
            <span class="s1">gc_collect_harder()</span>
            <span class="s1">record.append(</span><span class="s3">&quot;done&quot;</span><span class="s1">)</span>
        <span class="s0">else</span><span class="s1">:</span>
            <span class="s0">try</span><span class="s1">:</span>
                <span class="s1">token.run_sync_soon(collect_at_opportune_moment</span><span class="s0">, </span><span class="s1">token)</span>
            <span class="s0">except </span><span class="s1">_core.RunFinishedError:  </span><span class="s4"># pragma: no cover</span>
                <span class="s0">nonlocal </span><span class="s1">needs_retry</span>
                <span class="s1">needs_retry = </span><span class="s0">True</span>

    <span class="s0">async def </span><span class="s1">async_main():</span>
        <span class="s1">token = _core.current_trio_token()</span>
        <span class="s1">token.run_sync_soon(collect_at_opportune_moment</span><span class="s0">, </span><span class="s1">token)</span>
        <span class="s1">saved.append(agen())</span>
        <span class="s0">await </span><span class="s1">saved[-</span><span class="s2">1</span><span class="s1">].asend(</span><span class="s0">None</span><span class="s1">)</span>

    <span class="s4"># Actually running into the edge case requires that the run_sync_soon task</span>
    <span class="s4"># execute in between the system nursery's closure and the strong-ification</span>
    <span class="s4"># of runner.asyncgens. There's about a 25% chance that it doesn't</span>
    <span class="s4"># (if the run_sync_soon task runs before init on one tick and after init</span>
    <span class="s4"># on the next tick); if we try enough times, we can make the chance of</span>
    <span class="s4"># failure as small as we want.</span>
    <span class="s0">for </span><span class="s1">attempt </span><span class="s0">in </span><span class="s1">range(</span><span class="s2">50</span><span class="s1">):</span>
        <span class="s1">needs_retry = </span><span class="s0">False</span>
        <span class="s0">del </span><span class="s1">record[:]</span>
        <span class="s0">del </span><span class="s1">saved[:]</span>
        <span class="s1">_core.run(async_main)</span>
        <span class="s0">if </span><span class="s1">needs_retry:  </span><span class="s4"># pragma: no cover</span>
            <span class="s0">if not </span><span class="s1">buggy_pypy_asyncgens:</span>
                <span class="s0">assert </span><span class="s1">record == [</span><span class="s3">&quot;cleaned up&quot;</span><span class="s1">]</span>
        <span class="s0">else</span><span class="s1">:</span>
            <span class="s0">assert </span><span class="s1">record == [</span><span class="s3">&quot;final collection&quot;</span><span class="s0">, </span><span class="s3">&quot;done&quot;</span><span class="s0">, </span><span class="s3">&quot;cleaned up&quot;</span><span class="s1">]</span>
            <span class="s0">break</span>
    <span class="s0">else</span><span class="s1">:  </span><span class="s4"># pragma: no cover</span>
        <span class="s1">pytest.fail(</span>
            <span class="s3">f&quot;Didn't manage to hit the trailing_finalizer_asyncgens case &quot;</span>
            <span class="s3">f&quot;despite trying </span><span class="s0">{</span><span class="s1">attempt</span><span class="s0">} </span><span class="s3">times&quot;</span>
        <span class="s1">)</span>


<span class="s0">async def </span><span class="s1">step_outside_async_context(aiter):</span>
    <span class="s4"># abort_fns run outside of task context, at least if they're</span>
    <span class="s4"># triggered by a deadline expiry rather than a direct</span>
    <span class="s4"># cancellation.  Thus, an asyncgen first iterated inside one</span>
    <span class="s4"># will appear non-Trio, and since no other hooks were installed,</span>
    <span class="s4"># will use the last-ditch fallback handling (that tries to mimic</span>
    <span class="s4"># CPython's behavior with no hooks).</span>
    <span class="s4">#</span>
    <span class="s4"># NB: the strangeness with aiter being an attribute of abort_fn is</span>
    <span class="s4"># to make it as easy as possible to ensure we don't hang onto a</span>
    <span class="s4"># reference to aiter inside the guts of the run loop.</span>
    <span class="s0">def </span><span class="s1">abort_fn(_):</span>
        <span class="s0">with </span><span class="s1">pytest.raises(StopIteration</span><span class="s0">, </span><span class="s1">match=</span><span class="s3">&quot;42&quot;</span><span class="s1">):</span>
            <span class="s1">abort_fn.aiter.asend(</span><span class="s0">None</span><span class="s1">).send(</span><span class="s0">None</span><span class="s1">)</span>
        <span class="s0">del </span><span class="s1">abort_fn.aiter</span>
        <span class="s0">return </span><span class="s1">_core.Abort.SUCCEEDED</span>

    <span class="s1">abort_fn.aiter = aiter</span>

    <span class="s0">async with </span><span class="s1">_core.open_nursery() </span><span class="s0">as </span><span class="s1">nursery:</span>
        <span class="s1">nursery.start_soon(_core.wait_task_rescheduled</span><span class="s0">, </span><span class="s1">abort_fn)</span>
        <span class="s0">await </span><span class="s1">_core.wait_all_tasks_blocked()</span>
        <span class="s1">nursery.cancel_scope.deadline = _core.current_time()</span>


<span class="s1">@pytest.mark.skipif(buggy_pypy_asyncgens</span><span class="s0">, </span><span class="s1">reason=</span><span class="s3">&quot;pypy 7.2.0 is buggy&quot;</span><span class="s1">)</span>
<span class="s0">async def </span><span class="s1">test_fallback_when_no_hook_claims_it(capsys):</span>
    <span class="s0">async def </span><span class="s1">well_behaved():</span>
        <span class="s0">yield </span><span class="s2">42</span>

    <span class="s0">async def </span><span class="s1">yields_after_yield():</span>
        <span class="s0">with </span><span class="s1">pytest.raises(GeneratorExit):</span>
            <span class="s0">yield </span><span class="s2">42</span>
        <span class="s0">yield </span><span class="s2">100</span>

    <span class="s0">async def </span><span class="s1">awaits_after_yield():</span>
        <span class="s0">with </span><span class="s1">pytest.raises(GeneratorExit):</span>
            <span class="s0">yield </span><span class="s2">42</span>
        <span class="s0">await </span><span class="s1">_core.cancel_shielded_checkpoint()</span>

    <span class="s0">with </span><span class="s1">restore_unraisablehook():</span>
        <span class="s0">await </span><span class="s1">step_outside_async_context(well_behaved())</span>
        <span class="s1">gc_collect_harder()</span>
        <span class="s0">assert </span><span class="s1">capsys.readouterr().err == </span><span class="s3">&quot;&quot;</span>

        <span class="s0">await </span><span class="s1">step_outside_async_context(yields_after_yield())</span>
        <span class="s1">gc_collect_harder()</span>
        <span class="s0">assert </span><span class="s3">&quot;ignored GeneratorExit&quot; </span><span class="s0">in </span><span class="s1">capsys.readouterr().err</span>

        <span class="s0">await </span><span class="s1">step_outside_async_context(awaits_after_yield())</span>
        <span class="s1">gc_collect_harder()</span>
        <span class="s0">assert </span><span class="s3">&quot;awaited something during finalization&quot; </span><span class="s0">in </span><span class="s1">capsys.readouterr().err</span>


<span class="s1">@pytest.mark.skipif(buggy_pypy_asyncgens</span><span class="s0">, </span><span class="s1">reason=</span><span class="s3">&quot;pypy 7.2.0 is buggy&quot;</span><span class="s1">)</span>
<span class="s0">def </span><span class="s1">test_delegation_to_existing_hooks():</span>
    <span class="s1">record = []</span>

    <span class="s0">def </span><span class="s1">my_firstiter(agen):</span>
        <span class="s1">record.append(</span><span class="s3">&quot;firstiter &quot; </span><span class="s1">+ agen.ag_frame.f_locals[</span><span class="s3">&quot;arg&quot;</span><span class="s1">])</span>

    <span class="s0">def </span><span class="s1">my_finalizer(agen):</span>
        <span class="s1">record.append(</span><span class="s3">&quot;finalizer &quot; </span><span class="s1">+ agen.ag_frame.f_locals[</span><span class="s3">&quot;arg&quot;</span><span class="s1">])</span>

    <span class="s0">async def </span><span class="s1">example(arg):</span>
        <span class="s0">try</span><span class="s1">:</span>
            <span class="s0">yield </span><span class="s2">42</span>
        <span class="s0">finally</span><span class="s1">:</span>
            <span class="s0">with </span><span class="s1">pytest.raises(_core.Cancelled):</span>
                <span class="s0">await </span><span class="s1">_core.checkpoint()</span>
            <span class="s1">record.append(</span><span class="s3">&quot;trio collected &quot; </span><span class="s1">+ arg)</span>

    <span class="s0">async def </span><span class="s1">async_main():</span>
        <span class="s0">await </span><span class="s1">step_outside_async_context(example(</span><span class="s3">&quot;theirs&quot;</span><span class="s1">))</span>
        <span class="s0">assert </span><span class="s2">42 </span><span class="s1">== </span><span class="s0">await </span><span class="s1">example(</span><span class="s3">&quot;ours&quot;</span><span class="s1">).asend(</span><span class="s0">None</span><span class="s1">)</span>
        <span class="s1">gc_collect_harder()</span>
        <span class="s0">assert </span><span class="s1">record == [</span><span class="s3">&quot;firstiter theirs&quot;</span><span class="s0">, </span><span class="s3">&quot;finalizer theirs&quot;</span><span class="s1">]</span>
        <span class="s1">record[:] = []</span>
        <span class="s0">await </span><span class="s1">_core.wait_all_tasks_blocked()</span>
        <span class="s0">assert </span><span class="s1">record == [</span><span class="s3">&quot;trio collected ours&quot;</span><span class="s1">]</span>

    <span class="s0">with </span><span class="s1">restore_unraisablehook():</span>
        <span class="s1">old_hooks = sys.get_asyncgen_hooks()</span>
        <span class="s1">sys.set_asyncgen_hooks(my_firstiter</span><span class="s0">, </span><span class="s1">my_finalizer)</span>
        <span class="s0">try</span><span class="s1">:</span>
            <span class="s1">_core.run(async_main)</span>
        <span class="s0">finally</span><span class="s1">:</span>
            <span class="s0">assert </span><span class="s1">sys.get_asyncgen_hooks() == (my_firstiter</span><span class="s0">, </span><span class="s1">my_finalizer)</span>
            <span class="s1">sys.set_asyncgen_hooks(*old_hooks)</span>
</pre>
</body>
</html>